src/main/groovy/com/calgaryscientific/gradle/VeracodeDetailedReport.groovy

```diff
@@ -27,10 +27,46 @@
 package com.calgaryscientific.gradle
 
 import groovy.transform.CompileStatic
+import org.apache.commons.csv.CSVFormat
+import org.apache.commons.csv.CSVPrinter
 
 @CompileStatic
 class VeracodeDetailedReport {
 
+    static File getFile(String dir, String build_id) {
+        return new File(dir, "detailed-report-${build_id}.xml")
+    }
+
+    /**
+     * Extracts the software_composition_analysis information of the detailed XML report and return a list of rows with it
+     * @param xml - detailed report.
+     * @param file - file to write the report to.
+     *
+     * Component tree:
+     *
+     * /component @description @file_name @library @max_cvss_score @sha1 @vendor @version @vulnerabilities
+     *  /file_paths
+     *      /file_path @value
+     *  /vulnerabilities
+     *      /vulnerability @cve_id @cve_summary @cvss_score @cwe_id @severity
+     *  /violated_policy_rules
+     */
+    static List<List<String>> softwareCompositionAnalysisRows(Node xml) {
+        List<List<String>> rows = []
+        List<String> componentFields = ['library', 'file_name', 'vendor', 'description']
+        List<String> vulnerabilityFields = ['cve_id', 'cwe_id', 'cvss_score', 'severity', 'cve_summary']
+        // header row
+        rows.add(componentFields + vulnerabilityFields)
+        for (Node component : XMLIO.getNodeList(xml, 'software_composition_analysis', 'vulnerable_components', 'component')) {
```

Comment:
not sure if you would be interested in this groovy idiom
```
componentWithVulnerabilities = XMLIO.getNodeList(xml, 'software_composition_analysis', 'vulnerable_components', 'component').filter { it -> it.(component.attribute('vulnerabilities') as Integer) > 0 }
rowVulnerabilities = componentWithVulnerabilities.collect {  component -> XMLIO.getNodeList(component, 'vulnerabilities', 'vulnerability')).collect { [component, it]} }
result = rowVulnerabilities.collect { component, Vulnerability -> MLIO.getNodeAttributes(component, componentFields) + XMLIO.getNodeAttributes(vulnerability, vulnerabilityFields) }
```

Reply:
Thanks Tony, I refactored it with the `.each` idiom to be consistent with the rest of the codebase. Other than that, I don't think adding the filter and the extra lines adds to clarity.